### PR TITLE
Added option to add Jenkins init scripts

### PR DESCRIPTION
**Pull request by electroma**
In case you need configure Jenkins master:
* register global pipeline libraries
* set global properties
* bootstrap seed jobs

This PR adds option to pass a list of groovy snippets.
Looks a bit rough, but it is the only way I can see.

I've decided to inline scripts because generally they are just 2-3 lines long.
Btw, we could get rid of that xml blob and replace it with nice groovy snippet registering the cloud.

### CODE REVIEW

File: stable/jenkins/templates/jenkins-master-deployment.yaml [hunk 1]
```diff
@@ -27,7 +27,7 @@ spec:
                   "name": "copy-default-config",
                   "image": "{{.Values.Master.Image}}:{{.Values.Master.ImageTag}}",
                   "imagePullPolicy": "{{.Values.Master.ImagePullPolicy}}",
-                  "command": ["cp", "-n", "/var/jenkins_config/config.xml", "/var/jenkins_home"],
+                  "command": [ "bash", "/var/jenkins_config/apply_config.sh" ],
```
Comment (lachie83) [inline stable/jenkins/templates/jenkins-master-deployment.yaml]: Can we use `sh` instead of bash?

Comment (electroma) [reply stable/jenkins/templates/jenkins-master-deployment.yaml]: Fixed

**Comment by electroma** [comment]
Merged from master

**Comment by electroma** [comment]
@lachie83 looks like my change may not play well with custom ConfigMap.
I can move `apply_config.sh` out of ConfigMap or just add a note about it under "Custom ConfigMap".

What do you think?

**Comment by electroma** [comment]
@lachie83 any chance to have it merged?

**Comment by lachie83** [comment]
@electroma I think this should be good to go. Just doing some testing.

**Comment by lachie83** [comment]
@k8s-bot ok to test

**Comment by seanknox** [comment]
Hey @electroma, I'm getting an error from shell script created via the config map:

```
  12s		12s		1	{kubelet k8s-agent-265d0a06-0}							Warning		FailedSync	Error syncing pod, skipping: failed to "InitContainer" for "copy-default-config" with RunInitContainerError: "init container \"copy-default-config\" exited with 1"

  11s	10s	2	{kubelet k8s-agent-265d0a06-0}	spec.initContainers{copy-default-config}	Warning	BackOff		Back-off restarting failed docker container
  11s	10s	2	{kubelet k8s-agent-265d0a06-0}							Warning	FailedSync	Error syncing pod, skipping: [failed to "InitContainer" for "copy-default-config" with RunInitContainerError: "init container \"copy-default-config\" exited with 1"
, failed to "StartContainer" for "copy-default-config" with CrashLoopBackOff: "Back-off 10s restarting failed container=copy-default-config pod=handy-porcupine-jenkins-2874324452-j9m7k_default(3716c98f-0b63-11e7-b3e1-000d3a12bbe0)"
```

**Comment by electroma** [comment]
@seanknox yes, my bad.
It fails when there are no init scripts provided.
Let me fix it.

**Comment by electroma** [comment]
@seanknox fixed. Please try again.

**Comment by rossedman** [comment]
@electroma Do you mind posting an example of how this works?

**Comment by electroma** [comment]
@rossedman Here is example of configuration file including:
-seed job creation
-global settings
-pipeline library registration

https://github.com/buildit/digitalrig-docker/blob/master/helm/ec2_vars/jenkins.yaml

**Comment by rossedman** [comment]
@electroma Thank you! That makes sense and was the direction I was heading towards.

**Comment by plytro** [comment]
We can no longer rely on this as an upstream chart after this change. Our internal chart has a ConfigMap that includes credentials.xml, jenkins.model.JenkinsLocationConfiguration.xml, and custom config.xml. Since we provide a different config.yaml file the shell script will not be included and the chart will fail to install. Our goal is to not require forking this repo and just use it in requirements.yaml.

**Comment by electroma** [comment]
@plytro please provide an example of failing chart configuration

**Comment by plytro** [comment]
The only difference between these two charts is the version of stable chart I rely on.

https://github.com/plytro/jenkins-chart/tree/v0.2.0-working
https://gist.github.com/plytro/e410c7c596c550e58587a3e00c4efc9d
vs 
https://github.com/plytro/jenkins-chart/tree/v0.3.0-broken
https://gist.github.com/plytro/461c9b728fed02e4908754f8bbe5e60f

**Comment by rossedman** [comment]
@plytro From a quick glance you are using a custom config and it looks like the template for that has changed in this version. Could that be the problem? Also that seems to be what the output is on your error:

[This is where the changes are for this file](https://github.com/kubernetes/charts/commit/b889623d08158dce4ea9c06d146e80ec250d9379#diff-7146a5152507fba3404ff5118aab029c)

**Comment by plytro** [comment]
@rossedman Yes, I know that is exactly the problem. The chart has a configuration option for a custom configuration. With this change, if you set `CustomConfig: true`, then the shell script won't be included and the startup fails.

**Pull request merged**